cmd/gencerts.go

```diff
@@ -0,0 +1,147 @@
+package cmd
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"flag"
+	"math/big"
+	"net"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/OpenBazaar/openbazaar-go/repo"
+)
+
+//GenerateCertificates struct
+type GenerateCertificates struct {
+	DataDir  string        `short:"d" long:"datadir" description:"specify the data directory to be used"`
+	Testnet  bool          `short:"t" long:"testnet" description:"config file is for testnet node"`
+	Host     string        `short:"h" long:"host" description:"comma-separated hostnames and IPs to generate a certificate for"`
+	ValidFor time.Duration `long:"duration" description:"duration that certificate is valid for"`
+}
+
+//return PublicKey
+func publicKey(priv interface{}) interface{} {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &k.PublicKey
+	default:
+		return nil
+	}
+}
+
+func pemBlockForKey(priv interface{}) *pem.Block {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(k)}
+	default:
+		return nil
+	}
+}
+
+//Execute gencerts command
+func (x *GenerateCertificates) Execute(args []string) error {
+	// Set repo path
+	repoPath, err := repo.GetRepoPath(x.Testnet)
+	if err != nil {
+		return err
+	}
+	if x.DataDir != "" {
+		repoPath = x.DataDir
+	}
+
+	flag.Parse()
+
+	//Check if host entered
+	if len(x.Host) == 0 {
+		log.Fatalf("Missing required --host parameter")
+	}
+
+	// Set default duration
+	if x.ValidFor == 0 {
+		x.ValidFor = 365 * 24 * time.Hour
+	}
+
+	var priv interface{}
+
+	//Generate key
+	priv, err = rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		log.Fatalf("failed to generate private key: %s", err)
+	}
+
+	//Set creation date
+	var notBefore = time.Now()
+	notAfter := notBefore.Add(x.ValidFor)
+
+	//Crate serial nmuber
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		log.Fatalf("failed to generate serial number: %s", err)
+	}
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Organization: []string{"OpenBazaar"},
+		},
+		NotBefore: notBefore,
+		NotAfter:  notAfter,
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	//Check if host ip or dns name and count their quantity
+	hosts := strings.Split(x.Host, ",")
+	for _, h := range hosts {
+		if ip := net.ParseIP(h); ip != nil {
+			template.IPAddresses = append(template.IPAddresses, ip)
+		} else {
+			template.DNSNames = append(template.DNSNames, h)
+		}
+	}
+
+	template.IsCA = true
+	template.KeyUsage |= x509.KeyUsageCertSign
```

Comment:
Why are these floating down here like this? Why not add them to the `template` declaration on L89?